Change implementation of outStream for Central API Client functions

Ibaqu opened the pull request:
## Purpose
Add option to disable 'Printing to console' for Central API Client functions.

Inorder to fix the following issue :

> When the `CentralAPIClient` is pulling a module, it prints the success message (and progress) to stdout. Since language server and vscode language client communicates over the stdin/stdout using JSON RPC, above behavior cause non-json content to be written into stdout. This breaks the language server protocol and causes LS to not work there onward.

Fixes #31711

## Approach
> Describe how you are implementing the solutions along with the design details.

## Samples
> Provide high-level details about the samples related to this feature.

## Remarks
> List any other known issues, related PRs, TODO items, or any other notes related to the PR.

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

Ibaqu commented:
Push package and Pull package which makes use of outStream has been modified as such

A boolean `disableOutStream` has been added which will disable all prints to the console.

```
public void pushPackage(Path balaPath, String org, String name, String version, String supportedPlatform,
                            String ballerinaVersion, boolean disableOutStream)
                            
public void pullPackage(String org, String name, String version, Path packagePathInBalaCache,
                            String supportedPlatform, String ballerinaVersion, boolean isBuild,
                            boolean disableOutStream)                            
```

Encased all out.println() with a check to prevent it from being printed. Both functions are overloaded with the defaults as `false` to prevent other features from breaking

hevayo commented on cli/central-client/src/main/java/org/ballerinalang/central/client/Utils.java:
```diff
@@ -274,6 +280,23 @@ private static void writeAndHandleProgress(InputStream inputStream, FileOutputSt
+        } catch (IOException e) {
+            PrintStream printStream = System.out;
+
+            printStream.println(logFormatter.formatLog(fullPkgName + "pulling the package from central failed"));
```
At this level this should be an exception and can be handled at a higher level. Then the exception can be converted to a diagnostic at the Package resolution

hevayo commented on cli/central-client/src/main/java/org/ballerinalang/central/client/Utils.java:
```diff
@@ -274,6 +280,23 @@ private static void writeAndHandleProgress(InputStream inputStream, FileOutputSt
+
     /**
      * Create nightly build meta file.
      *
```
Can we add some tests how can you disable printing when using ProjectAPI ?

Ibaqu replied:
Added tests. Will add tests to distribution as well.

azinneera commented on cli/central-client/src/main/java/org/ballerinalang/central/client/Utils.java:
```diff
@@ -274,6 +278,17 @@ private static void writeAndHandleProgress(InputStream inputStream, FileOutputSt
+        byte[] buffer = new byte[1024];
+
+        while ((count = inputStream.read(buffer)) > 0) {
+            outputStream.write(buffer, 0, count);
```
We should not write to the output stream. The idea is that we do the pull operation quietly. If the pull fails due to any interruption, then we should throw an exception and log it as a diagnostic.

Ibaqu commented:
Changes : 

- Change `DISABLE_OUTPUT_STREAM` -> `ENABLE_OUTPUT_STREAM` 
- Added `System.setProperty(CentralClientConstants.ENABLE_OUTPUT_STREAM, "false");` to `CompileTask` disabling output stream for all commands
- Added tests

azinneera commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CompileTask.java:
```diff
@@ -61,6 +62,8 @@ public void execute(Project project) {
         // Print the source
         this.out.println("\t" + sourceName);
 
+        System.setProperty(CentralClientConstants.ENABLE_OUTPUT_STREAM, "false");
```
The progress bar should show in the stdout in the case of commands.

Ibaqu replied:
Changed to true as discussed

azinneera commented on cli/central-client/src/main/java/org/ballerinalang/central/client/CentralAPIClient.java:
```diff
@@ -342,8 +347,14 @@ public void pushPackage(Path balaPath, String org, String name, String version,
     }
 
+    /**
+     * Pulling a package.
```
Shall we refactor the method comment to use proper tense and add the parameters and exceptions thrown?

Ibaqu replied:
Addressed in https://github.com/ballerina-platform/ballerina-lang/commit/a26d8a97e757d634eba17907656536dfe134e2d5

azinneera commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CompileTask.java:
```diff
@@ -61,6 +62,8 @@ public void execute(Project project) {
         // Print the source
         this.out.println("\t" + sourceName);
 
+        System.setProperty(CentralClientConstants.ENABLE_OUTPUT_STREAM, "true");
```
We need to set this in `PullCommand` and `PushCommand` as well

azinneera commented on cli/central-client/src/test/java/org/ballerinalang/central/client/TestCentralApiClient.java:
```diff
@@ -526,4 +527,91 @@ private void setBallerinaHome() {
         }
     }
+
+    @Test(description = "Test pull package without Logs", enabled = false)
```
We can move these tests to distribution without having it disabled here.

Ibaqu replied:
Removed disabled tests

Ibaqu commented:
Test : https://github.com/ballerina-platform/ballerina-distribution/pull/2070
